consensus/test/member_management_test.go

```diff
@@ -0,0 +1,210 @@
+  addValidatorHook, it issues transaction to call autonity contract via operator account to add a new validator.
+
+  Then the test case verify the output from its finalAssert hook function on the specified height of the blockchain, for
+  example the addValidatorCheckerHook checks if the new validator is presented in the white list, and its stake balance
+  checked too, and finally it checks the total stake supply after the membership updates.
+
+  for the other cases in the file: add stake_holder, participants, or remove user, they follow the same work flow and
+  some rules to check the outputs.
+*/
+
+func TestMemberManagement(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping test in short mode")
+	}
+	// prepare chain operator
+	operatorKey, err := crypto.GenerateKey()
+	if err != nil {
+		t.Fatal(err)
+	}
+	operatorAddress := crypto.PubkeyToAddress(operatorKey.PublicKey)
+
+	newValidatorNodeKey, err := crypto.GenerateKey()
+	require.NoError(t, err)
+	newValidatorPubKey := newValidatorNodeKey.PublicKey
+	newValidatorENode := enode.V4DNSUrl(newValidatorPubKey, "VN:8527", 8527, 8527)
+
+	newStakeholderKey, err := crypto.GenerateKey()
+	require.NoError(t, err)
+	newStakeholderPubKey := newStakeholderKey.PublicKey
+	newStakeholderEnode := enode.V4DNSUrl(newStakeholderPubKey, "SN:8528", 8528, 8528)
+
+	newParticipantKey, err := crypto.GenerateKey()
+	require.NoError(t, err)
+	newParticipantPubKey := newParticipantKey.PublicKey
+	newParticipantEnode := enode.V4DNSUrl(newParticipantPubKey, "PN:8529", 8529, 8529)
+
+	removeNodeKey, err := crypto.GenerateKey()
+	require.NoError(t, err)
+	removeNodePubKey := removeNodeKey.PublicKey
+	eNodeToRemove := enode.V4DNSUrl(removeNodePubKey, "VM:8527", 8527, 8527)
+	addressToRemove := crypto.PubkeyToAddress(removeNodePubKey)
+
+	validatorStake := new(big.Int).SetUint64(200)
+	stakeHolderStake := new(big.Int).SetUint64(100)
+	participantStake := new(big.Int).SetUint64(0)
+
+	// genesis hook
+	genesisHook := func(g *core.Genesis) *core.Genesis {
+		g.Config.AutonityContractConfig.Operator = operatorAddress
+		g.Alloc[operatorAddress] = core.GenesisAccount{
+			Balance: big.NewInt(100000000000000000),
+		}
+
+		// the user to be removed.
+		user := &params.User{
+			Address: &addressToRemove,
+			Enode:   eNodeToRemove,
+			Type:    "participant",
+			Stake:   0,
+		}
+		g.Config.AutonityContractConfig.Users = append(g.Config.AutonityContractConfig.Users, *user)
+		return g
+	}
+
+	addUser := func(operator *ecdsa.PrivateKey, port int, userPubKey ecdsa.PublicKey, stake *big.Int, enode string, userType uint8) error {
+		return interact(port).tx(operator).addUser(crypto.PubkeyToAddress(userPubKey), stake, enode, userType)
+	}
+
+	addUserHook := func(validator *testNode, _ common.Address, _ common.Address) (bool, *types.Transaction, error) { //nolint
+		if validator.lastBlock == 4 {
+			err := addUser(operatorKey, validator.rpcPort, newValidatorPubKey, validatorStake, newValidatorENode, uint8(2))
```

Comment:
It would be nice to assign the roles to variables, so we don't have these mystery values being duplicated in the test.

for example
```suggestion
			err := addUser(operatorKey, validator.rpcPort, newValidatorPubKey, validatorStake, newValidatorENode, validatorRole)
```

Reply:
> It would be nice to assign the roles to variables, so we don't have these mystery values being duplicated in the test.
> 
> for example

Resloved.